### PR TITLE
feat(migration): Delete alert rules with Alert Rule Actions

### PR DESCRIPTION
**Pull request by NisanthanNanthakumar**
## Objective:
Required for https://github.com/getsentry/sentry/pull/31381

We are updating how we store the settings data for Alert Rule Action UI Components. However during testing, we have some alert rules using our current (outdated) data structure. This migration will be used to delete all the `sentry_alertruletriggeractions` and `sentry_rule` records that have an `Alert Rule Action` UI Component

## Impact
https://redash.getsentry.net/queries/2440/source

11 rows will be deleted from `sentry_alertruletriggeractions`. All will be from Sentry or partner test apps.
We cannot determine how many `sentry_rule` records will be deleted bc the values are pickled. However only the rules from our test orgs will be affected.

### CODE REVIEW

File: src/sentry/migrations/0269_delete_legacy_alert_rule_action_records.py [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+    AlertRuleTriggerAction = apps.get_model("sentry", "AlertRuleTriggerAction")
+    SentryAppComponent = apps.get_model("sentry", "SentryAppComponent")
+
+    for trigger_action in AlertRuleTriggerAction.objects.filter(
+        sentry_app_id__in=SentryAppComponent.objects.filter(type="alert-rule-action").values_list(
+            "sentry_app_id", flat=True
+        ),
+        sentry_app_config__isnull=False,
+    ):
+        trigger_action.delete()
```
Comment (NisanthanNanthakumar) [inline src/sentry/migrations/0269_delete_legacy_alert_rule_action_records.py:25]: I think its ok to just delete the action and not the entire metric alert here.

File: src/sentry/migrations/0269_delete_legacy_alert_rule_action_records.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    # the most part, this should only be used for operations where it's safe to run the migration
+    # after your code has deployed. So this should not be used for most operations that alter the
+    # schema of a table.
+    # Here are some things that make sense to mark as dangerous:
+    # - Large data migrations. Typically we want these to be run manually by ops so that they can
+    #   be monitored and not block the deploy for a long period of time while they run.
+    # - Adding indexes to large tables. Since this can take a long time, we'd generally prefer to
+    #   have ops run this and not block the deploy. Note that while adding an index is a schema
+    #   change, it's completely safe to run the operation after the code has deployed.
+    is_dangerous = False
```
Comment (NisanthanNanthakumar) [inline src/sentry/migrations/0269_delete_legacy_alert_rule_action_records.py:68]: Im debating marking this dangerous. However, I want to make the migration work for onprem users.

Comment (markstory) [reply src/sentry/migrations/0269_delete_legacy_alert_rule_action_records.py:68]: self-hosted and single-tenant ignore the `is_dangerous` flag and run all migrations.

File: src/sentry/migrations/0269_delete_legacy_alert_rule_action_records.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+
+def delete_metric_alert_trigger_actions_with_alert_rule_actions(apps, schema_editor):
+    """
+    Delete the rows in the AlertRuleTriggerAction table that use an Alert Rule Action SentryApp.
+    """
+    AlertRuleTriggerAction = apps.get_model("sentry", "AlertRuleTriggerAction")
+    SentryAppComponent = apps.get_model("sentry", "SentryAppComponent")
+
+    for trigger_action in AlertRuleTriggerAction.objects.filter(
```
Comment (mgaeta) [inline src/sentry/migrations/0269_delete_legacy_alert_rule_action_records.py:19]: I think there are two ways to do this query more efficiently:
```py
query = AlertRuleTriggerAction.objects.filter(
    sentry_app__type="alert-rule-action",
    sentry_app_config__isnull=False,
)

for trigger_action in query:
    ...
```
```py
query = AlertRuleTriggerAction.objects.all().select_related("sentry_app")
for alert_rule_trigger_action in RangeQuerySetWrapperWithProgressBar(query):
    if (
        alert_rule_trigger_action.sentry_app.type == "alert-rule-action" 
        and alert_rule_trigger_action.sentry_app_config is not None
    ):

```
You'd have to try them out though I haven't tested this.

**Review by wedamija — COMMENTED** [review]
If we're just converting from camelCase to snake case would it make more sense to convert these rows rather than delete them?

What's the context for how these are set up? Did we set them up as a test or are users beta testing?

Feels a little weird to delete these actions but not the entire rules. Won't this leave some rules with no actions?

File: src/sentry/migrations/0269_delete_legacy_alert_rule_action_records.py [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+    """
+    SentryAppComponent = apps.get_model("sentry", "SentryAppComponent")
+    SentryAppInstallation = apps.get_model("sentry", "SentryAppInstallation")
+    Project = apps.get_model("sentry", "Project")
+    Rule = apps.get("sentry", "Rule")
+
+    # First find all projects by orgs that have the sentryapps with Alert Rule Actions installed
+    # Filter sentry_rule table by the affected projects.
+    # Unpickle each rule to determine if they have an Alert Rule Action and then delete
+    for rule in RangeQuerySetWrapperWithProgressBar(
```
Comment (mgaeta) [inline src/sentry/migrations/0269_delete_legacy_alert_rule_action_records.py:40]: I think making this a single query would look like 
```py
query = Rule.objects.filter(
    project__organization__sentry_app_installations__sentry_app__components__type="alert-rule-action"
)
for rule in query:
    ...
```
or 
```py
for rule in RangeQuerySetWrapperWithProgressBar(Rule.objects.all()):
    installs = rule.project.organization.sentry_apps.sentry_app_installations.all()
    if any([install.sentry_app.components.type == "alert-rule-action" for install in installs]):
        ...
```

**Comment by NisanthanNanthakumar** [comment]
> If we're just converting from camelCase to snake case would it make more sense to convert these rows rather than delete them?
>
> What's the context for how these are set up? Did we set them up as a test or are users beta testing?

@wedamija it'll be whatever case that is defined by the SentryApp's schema for Alert Rule Action UI Component. From the Redash query, the only orgs that setup and installed an SentryApp with an Alert Rule Action is ourselves and Linear's dev org. So all rules created were for testing.

> Feels a little weird to delete these actions but not the entire rules. Won't this leave some rules with no actions?
@wedamija Hmm yea but we can currently save Metric Alerts without an action. It wont matter because all the orgs are just test orgs.

File: src/sentry/migrations/0269_delete_legacy_alert_rule_action_records.py (hunk 3 above)
Comment (NisanthanNanthakumar) [reply src/sentry/migrations/0269_delete_legacy_alert_rule_action_records.py:19]: @mgaeta the `sentry_app_id` column in `sentry_alertruletriggeraction` is a FK to `sentry_sentryapp` not `sentry_sentryappcomponent`. `type` is stored in `sentry_sentryappcomponent`

**Comment by wedamija** [comment]
@NisanthanNanthakumar If these only exist for these test orgs, and not in on-prem etc, should we just go through manually and delete these via the UI? It looks like there are 11 rows, not sure it's actually worth the time to automate this.

**Comment by NisanthanNanthakumar** [comment]
> @NisanthanNanthakumar If these only exist for these test orgs, and not in on-prem etc, should we just go through manually and delete these via the UI? It looks like there are 11 rows, not sure it's actually worth the time to automate this.

@wedamija haha I didn't know that was an option! hmm makes sense for the metric alerts but for Issue Alerts since the data column is pickled, I can only scope down to 182 records (the last query in the Redash file) that need to be inspected for the Alert Rule Action. I highly doubt all of them have Alert Rule Actions. Maybe we can have someone from Ops run a python script to unpickle and return the ids of the Issue Alerts that need to be deleted?

File: src/sentry/migrations/0269_delete_legacy_alert_rule_action_records.py (hunk 4 above)
Comment (NisanthanNanthakumar) [reply src/sentry/migrations/0269_delete_legacy_alert_rule_action_records.py:40]: @mgaeta The first query is just doing inner joins. It's not as good as the original query for reducing the number of rows to iterate over. Outputs:
```
SELECT sentry_rule.id, sentry_rule.project_id, sentry_rule.environment_id, sentry_rule.label, sentry_rule.data, sentry_rule.status, sentry_rule.owner_id, sentry_rule.date_added
        FROM sentry_rule
        INNER JOIN sentry_project ON
    (sentry_rule.project_id = sentry_project.id)
    INNER JOIN sentry_organization ON
    (sentry_project.organization_id = sentry_organization.id)
    INNER JOIN sentry_sentryappinstallation ON
    (sentry_organization.id = sentry_sentryappinstallation.organization_id) INNER JOIN sentry_sentryapp
    ON (sentry_sentryappinstallation.sentry_app_id = sentry_sentryapp.id) INNER JOIN sentry_sentryappcomponent
    ON (sentry_sentryapp.id = sentry_sentryappcomponent.sentry_app_id) WHERE sentry_sentryappcomponent.type = 'alert-rule-action'
```
1563 rows

Original Query: 182 rows

The second option looks like it'll loop over all the records in `sentry_rule` which 9623311 rows. And also FK relations are the other way around. Rule -> Project -> Organization <- Sentry_App_Installation <- Sentry_App <- Sentry_App_Component

**Comment by wedamija** [comment]
> 



> > @NisanthanNanthakumar If these only exist for these test orgs, and not in on-prem etc, should we just go through manually and delete these via the UI? It looks like there are 11 rows, not sure it's actually worth the time to automate this.
> 
> @wedamija haha I didn't know that was an option! hmm makes sense for the metric alerts but for Issue Alerts since the data column is pickled, I can only scope down to 182 records (the last query in the Redash file) that need to be inspected for the Alert Rule Action. I highly doubt all of them have Alert Rule Actions. Maybe we can have someone from Ops run a python script to unpickle and return the ids of the Issue Alerts that need to be deleted?

Oh I thought there were just 11 rows you could manually delete by looking at these test orgs, if there are a lot more than this is fine. There used to be a table that had these decoded for issue alert rules in bigquery, you could ask data if it still exists.

**Comment by NisanthanNanthakumar** [comment]
> Oh I thought there were just 11 rows you could manually delete by looking at these test orgs, if there are a lot more than this is fine. There used to be a table that had these decoded for issue alert rules in bigquery, you could ask data if it still exists.

@wedamija 👍  I'll look into that! I remember there was an issue with bigquery timing out trying to unpickle the records a couple months ago. Hopefully that's solved!

**Comment by NisanthanNanthakumar** [comment]
@wedamija 
So for Issue Alerts: we have 7 rules that need to be deleted 
https://redash.getsentry.net/queries/2445/source

And for Metric Alerts we have 10 actions to be deleted.
https://redash.getsentry.net/queries/2440/source

Since this is doable manually, we can close this PR. Before deleting I'll send out an email to the partner org that we will be deleting their test rules.

**Pull request closed without merging**